### DISCUSSION TITLE
conditional fetch returns undefined between renders

### DISCUSSION THREAD

**Discussion by lucksp** [vercel/swr#1490] [Q&A] [answered]
Based on [the docs for conditional fetching](https://swr.vercel.app/docs/conditional-fetching#conditional) it's easy to see how to implement this functionality.  However, we are running into a situation that is unexpected, or at least unclear of intended behavior related to the docs line:
> SWR will not start the request.

We know SWR uses the URL to base its caching and therefore the ability to determine whether or not the `fetcher` should be invoked.

Currently we build a hook like:
```
const {
    items,
    isLoading: isFleetItemsLoading,
  } = useFleetData({
    orgId,
    boundingBox,
    shouldFetch:
      !isMapLoaded || // map hasn't loaded; make initial call
      (isMapLoaded && state.isSearchable), // `Search Area` button called; make call with bounding box
  });
```
and then:
```
let urlKey = `api/route/${orgId}`;
  if (boundingBox) {
        urlKey = `${urlKey}?addMoreParamsHere`;
  }

const {data} = useSWR(shouldFetch ? [urlKey, user] : null,  (url, auth) => get(url, { token: auth.token }));
```

So on the first load, `shouldFetch` is true because `!isMapLoaded` is true so we query with a url without params.  We **successfully** get data back.

Then the `shouldFetch` goes to `false`...

what happens here is that data becomes `undefined`. 
![image](https://user-images.githubusercontent.com/15162169/134382045-cf159c06-4389-4bab-af07-e569ab80d00c.png)

The question is:  if there is `data` and the conditional logic says "do not fetch" should data be reset or keep its prior state?
The documentation implies it would not reset anything...
yet I can picture why this happens, because no url === undefined data.

**Comment by shuding** [comment, marked as answer]
Yeah that's the expected behavior, because in SWR the key is the identifier of the resource: when the key becomes undefined, data will become undefined too, this ensures the correctness of SWR (otherwise there might be inconsistency in the UI, imagine if both key and data are displayed on the same screen).

But sometimes it makes sense to keep the previous result and make the hook "laggy", the recommended solution for now is mentioned here: https://swr.vercel.app/docs/middleware#keep-previous-result.
